### ISSUE TITLE
Enhancement: Discuss the debug facility documentation and code so it can be split to another repo

### ISSUE THREAD

**Issue by xexyl** [ioccc-src/mkiocccentry#232]
Rather than clutter up issue #171 I think we can discuss the concerns here. As I want to go afk again (in fact need to go afk again) I will not copy over the comments regarding it. If you've not replied to any and I have the time (and not working on the actual changes) maybe I will copy them over here.

Going afk a while ... not sure when I'll be back. Feel free to assign this issue to me - and to you if you wish.

**Comment by xexyl** [comment]
Update: I'm feeling really tired right now so I'm going to have to do something else for now. I hope to get back to it later on today. In any event since it's a lot to go through it'll probably take a few days. I think the next time I'm at it I'll see if I can find a good place to commit and then do that. Then go from there.

Since I'm tired perhaps I should take the time to rest until the questions I asked in the other thread have been answered/discussed. For now though I'm done. I do have many changes and the most important one I already did a pull request for - the bug fix `-h` not printing anything at all.

So either more work will be done today on this or else I'll continue tomorrow morning. I somehow suspect that that'll be when there's another commit from me but we shall see. Good luck on the Internet connectivity issue (and maybe the fact you're having that problem will make it slower in resolving/discussing this issue as well). 

Whatever the case I'm making good progress! I got through some 500 lines in the markdown file but I want to go over those lines again. I've noticed some things that are throughout the file (beyond the line I got to) so I fixed those as well - then I want to make sure everything reads okay. I also added some clarifications to comments in `dbg.c`. I think it'll look quite good once it's complete though.

Good day!

**Comment by xexyl** [comment]
Just a quick note. 

This morning I had something else I was working on but tomorrow I shouldn’t have that problem. I therefore hope I can get more done tomorrow than I did today. 

I am still thinking about the best way to go about it. I might make some quick passes of text that has very similar if not the same text. 

What I do know is that I will have to make several passes over the text. Some of it is also in `dbg.c` as well. For example the comments of some functions. 

To make it less error prone I will simply do the replacement in two different vim windows (as in `:new file` and `:wincmd …`). Some of this is already done but I will go over them all again. 

I will be making several commits and it’s likely that after I make some and open a pull request that I will make more commits. As we discuss the different things that come up it’s quite likely that more commits will be necessary. 

Hopefully we can get this finished by the end of the week. I would say that (and this is typically the model I use) after I have made what I think is the last commit I will probably go over it one more time just to be sure nothing else is noticed during the final read. 

I think that’s all that is on my mind wrt this at this. I will be awake for a while but then will be going to sleep. Hopefully sleep won’t be as difficult as it was last night but if it is hopefully it won’t affect me that much. 

Good night!

**Comment by xexyl** [comment]
Making good progress but I came across the `warn_or_err` function and it reminded me: there seems to be a really odd problem with it. I'm not sure exactly what happens but it seems to be buggy anyway.

I tried using it in earlier versions of the json check function (for the checks that are done in both files) and I got really strange output and invalid check results as well. I might be able to find it in mail but if not it might be a function that should be tested.

Other things I'm doing:

Where possible I've moved some of the same messages to a general notes section at the top of the file. For example when `fmt == NULL`. There are places where this is not possible as it depends on other variables but this type of thing should make for shorter reading and also make it easier to fix any inconsistencies/errors/whatever.

I'm probably going to take a break soonish. I will make a pull request later on I'm sure.

**Comment by xexyl** [comment]
Oh as for `warn_or_err` and the problem: I just tried it and it seems to not be there. But it might also be masked by what's allowed to be seen and not. I'm not sure. It also might have been fixed.

However I still think there might be another problem: not sure what you think here. As it stands if it's an error and error messages are not allowed to be displayed it returns; it does not exit. Is this how it should be though? To me that's like saying it's not an error based on a boolean that has nothing to do with the actual error condition.

I'm not sure if other functions have this same problem (fault?) but if you want me to update it I can.

As for the other problem I'm not sure - I have to do more checks and it might be the above problem is masking it. Let me know what you think.

Really off to do something else now. Good day!

**Comment by lcn2** [comment]
> Oh as for `warn_or_err` and the problem: I just tried it and it seems to not be there. But it might also be masked by what's allowed to be seen and not. I'm not sure. It also might have been fixed.
> 
> However I still think there might be another problem: not sure what you think here. As it stands if it's an error and error messages are not allowed to be displayed it returns; it does not exit. Is this how it should be though? To me that's like saying it's not an error based on a boolean that has nothing to do with the actual error condition.
> 
> I'm not sure if other functions have this same problem (fault?) but if you want me to update it I can.
> 
> As for the other problem I'm not sure - I have to do more checks and it might be the above problem is masking it. Let me know what you think.
> 
> Really off to do something else now. Good day!
> 
> Quick update:
> 
> Took a look at the other functions and they just exit() if error output is not allowed. This makes me think that this check:
> 
> ```c
>     if ((warning == true &&
>             (warn_output_allowed == false ||
>              (msg_warn_silent == true && verbosity_level <= 0))) ||
>         (warning == false &&
>             err_output_allowed == true)) {
>         return;
>     }
> ```
> 
> is in error itself. It should do instead:
> 
> * If it's a warning and warning output is not allowed then return.
> * If it's an error though go through the rest of the function but don't print the error message if output is not allowed. After that just exit so that it is similar to the other error functions.
> 
> What do you think?
> 
> Oh and when looking at the code itself I noticed that it returns if error output is allowed rather than not allowed. If nothing else that should be fixed but I'll wait for your reply to find out what you think should be done in full. Now I really am off to do something else.

You are correct in that `warn_or_err()` should exit when `warning == false` and output is disabled.

See 9883db473b005c7bc53bba58ea85561c1ce732cb

    Fix warn_or_err to exit when warning is false and output disabled

**Comment by lcn2** [comment]
We very much appreciate your working on the _dbg_ facility.

Let us know when you believe you are done.

**Comment by xexyl** [comment]
> We very much appreciate your working on the _dbg_ facility.
> 

I consider it an incredible honour and privilege that you asked for my help! Thank you!

> Let us know when you believe you are done.

I certainly will my friend!

I had hoped to get it done by the end of the week but that’s not likely. Tomorrow I hope to make good progress but I am not sure how much that will be. I will be going to sleep soon (though it always takes a while before I actually fall asleep). 

I think I would not have managed to get what I did today if I hadn’t managed to sleep in a while. It was a difficult night. 

This next week I have three doctor appointments so I will probably not be able to do as much later in the day but that’s never certain. 

I will look at the other threads briefly before I go get ready to sleep but hopefully tomorrow I can make good progress on this!

**Comment by lcn2** [comment]
Hope you rest better this evening, @xexyl

**Comment by xexyl** [comment]
> Hope you rest better this evening, @xexyl

Thank you. Unfortunately I didn’t. Had some awful dreams. I woke up quite early again but I am hoping I can get some documentation done for `dbg` in a bit.

**Comment by xexyl** [comment]
I was thinking of updating the file but I think it might be better to first discuss some things rather than make presumptions (or even worse assumptions) on what you want. There are some things that can be done no matter what: typo fixes for example.

But the rest of it is less clear. So I think what I'll do is go through the document (I will try now but it might be slow as I'm quite tired) and when reading I'll stop and think about each section (and any subsections as well) and write comments here in a different comment for each. Then after discussing all of them (as some of them are connected) I can make the changes.

**Comment by xexyl** [comment]
## Hyperlinks

As I suggested the other day I think this would help a lot. But what sort of hyperlinks? Obviously a table of contents. I would think that each section would also have a short list of links as well: for example at the top of each section it would allow one to jump to the respective function.

Another thought: should each reference to a function (outside of code blocks) be a link to the function prototype? I could see how this might be an issue but it might also be helpful to some people.

What do you think? If I think of other things with links when going through the document I'll bring them up in another comment.

**Comment by xexyl** [comment]
## TL; DR section

I moved the error details to this section that were mostly duplicated across all the sections. There was an exception though where it depended on a boolean: that one I left there. I guess that was `warn_or_err` but maybe not.

As of commit https://github.com/ioccc-src/mkiocccentry/pull/240/commits/249f72541c04dfe11671a27a904d50f525ffd8c6 I did the same for printf formats. I think this type of thing will help make it easier to read. There might be some cases where this is not the case though including ironically if someone skips past the `tl;dr` section and goes straight to the functions they're after.

This makes me wonder if there should be a general synopsis section at the top as well. It might discuss the different capabilities of the facility and also have all the function prototypes (perhaps without the comments at the top). Then each prototype could be a link to the actual function details. Perhaps a back link would allow them to go back to this section.

That might make it easier to find the different facilities.

What do you think here?

**Comment by xexyl** [comment]
## Examples at the top of the file

Would it be beneficial to have at the top of the file (maybe in the general function section I suggested or brought up in https://github.com/ioccc-src/mkiocccentry/issues/232#issuecomment-1146792747 ?) an example which has all functions in use? I'm not sure: it might be too much for some people but it might also give an entire overview of the facility for people to see.

I'm not really sure here.

**Comment by xexyl** [comment]
## Synopses

If we do have the synopses of all functions at the top though what do we do about each section? We could remove the synopses there. But it might be a problem to some. If we did that we might also have to add the comments at the top of the functions to the synopses at the top.

On the one hand it would make for less clutter but on the other it could be harder for some people (and easier for others).

There's another way to consider though. It might be that since the naming scheme is consistent we could describe the naming scheme at the top, giving the suffix (name) of each facility. This is not what I propose here (it's quickly written and without much though - and a tired head) but it might give you the idea (and I don't know what I think of this either):

## Naming scheme of the facilities

The `dbg` facility provides functions for different types of informative messages: info, debug, error and warning messages. To make it easier to know which functions to use we provide a consistent naming scheme.

The functions that start with `f` write to a stream whereas those without write to `stderr`. 

### Informative messages: `msg()` and `fmsg()`

For informative messages this means that `msg` writes to `stderr` and `fmsg` writes to a `FILE *`.

--

And so on. Again **the example is NOT what I recommend! I'm not even sure if this is ideal: It just is an idea that popped into my head that _might_ make it easier for some. I hope you get the idea here though.

I'll have to be taking a break soon as writing that not-to-be-used idea took more effort than I'd like to believe! The comments above and this one might be a good starting point however. I'll take a look real quick to see if there's anything else that comes to my mind immediately and if so I'll write them but if not I'll continue looking at it later on - hopefully today.

**Comment by xexyl** [comment]
## Describing the functions on a whole

Perhaps this might be at the top with the synopses or the list of functions: not sure. But it could take the first subsection of each section and combine them. For example (yes it's out of order: doing this quickly):

# List of facilities

The `dbg` code allows you to write different types of messages: informative, debug, error and warning messages. Please see the following subsections for each type.

## dbg - write a verbosity level allowed debug message

The `dbg()` function will write a debug message, if the verbosity level is high
enough, to a stream, such as `stderr`.

The `fdbg()` function will write a debug message, if the verbosity level is high
enough, to a stream.

## msg - write a generic message

The `msg()` function will write an informational message to `stderr`.

The `fmsg()` function will write an informational message to a stream.

...

What to do about what follows each? I refer to:

```
## Synopsis:

```c
/*
 * msg - write a generic message, to stderr
 *
 * given:
 *      fmt     printf format
 *      ...
 */
extern void msg(const char *fmt, ...);

/*
 * fmsg - write a generic message, to a stream
 *
 * given:
 *      stream  open stream to use
 *      fmt     printf format
 *      ...
 */
extern void fmsg(FILE *stream, char const *fmt, ...);
```

Special care is made in these functions to restore the caller's original value
of `errno` on return.
```

Well the special care message is duplicated so possibly it could be moved to general notes. But I was more specifically referring to:

```
/*
 * msg - write a generic message, to stderr
 *
 * given:
 *      fmt     printf format
 *      ...
 */
```

and

```
/*
 * fmsg - write a generic message, to a stream
 *
 * given:
 *      stream  open stream to use
 *      fmt     printf format
 *      ...
 */
```

Well it might be that those should be in the synopses section as well. Then under each section (which would no longer have the above) it would have the examples. For instance under the message one:

```
## Examples:

```c
msg("This is an informative message.");
msg("foo: %d bar: <%s> baz: %g", foo, bar, baz);

/* ... */

fmsg(stdout, "This is an informative message to an open file: %s", debug_filename);
```

A newline (i.e., `\n`) is written after the message and thus the `fmt` argument
does **NOT** need to end in a newline (though it can if you want it to).

```

**Again these are VERY loose ideas!** I also am sorry for anything that might not be worded so well or not as clear. I'm going to take a break real soon and I hope that with some discussion I can start working on this tomorrow.

**Comment by xexyl** [comment]
One more and then I am taking a break for a while. (Make that two: we might want to bring up the `va_list` versions briefly at the top but point out that they're not recommended - for the reasons given).

## Output control

It might be an idea to move this towards the top too? It would kind of be a merge of the different ones. It might be in the synopses of each?

Might I also suggest we write (I am happy to do this) a man page? That could be a way to make it easier to see how to use the functions. Would it be section 3? I'd guess that but maybe you have another idea?

Going to take a break now. Will discuss these things with you and go from there. Again sorry for any inconsistencies or for any text that's not clear and even the poor examples. I just wanted to get the ideas out to discuss quickly so I could go rest.

**Comment by lcn2** [comment]
> ## Hyperlinks
> As I suggested the other day I think this would help a lot. But what sort of hyperlinks? Obviously a table of contents. I would think that each section would also have a short list of links as well: for example at the top of each section it would allow one to jump to the respective function.
> 
> Another thought: should each reference to a function (outside of code blocks) be a link to the function prototype? I could see how this might be an issue but it might also be helpful to some people.
> 
> What do you think? If I think of other things with links when going through the document I'll bring them up in another comment.

Probably a compact table of contents so that someone can quickly jump down into the middle of the document would be the most useful.

**Comment by lcn2** [comment]
> ## TL; DR section
> I moved the error details to this section that were mostly duplicated across all the sections. There was an exception though where it depended on a boolean: that one I left there. I guess that was `warn_or_err` but maybe not.
> 
> As of commit [249f725](https://github.com/ioccc-src/mkiocccentry/commit/249f72541c04dfe11671a27a904d50f525ffd8c6) I did the same for printf formats. I think this type of thing will help make it easier to read. There might be some cases where this is not the case though including ironically if someone skips past the `tl;dr` section and goes straight to the functions they're after.
> 
> This makes me wonder if there should be a general synopsis section at the top as well. It might discuss the different capabilities of the facility and also have all the function prototypes (perhaps without the comments at the top). Then each prototype could be a link to the actual function details. Perhaps a back link would allow them to go back to this section.
> 
> That might make it easier to find the different facilities.
> 
> What do you think here?

What is moved into that section should be kept to a minimum as  making the `TL;DR` too long kind of defeats the purpose.  :-)

Moreover the `TL;DR` section should be a very short summary.  Perhaps it is better for the `TL;DR` section might refer to some _common error section_ and simply and briefly mention that if the user runs into problems with a link to that _common error section_?

**Comment by lcn2** [comment]
> ## Examples at the top of the file
> Would it be beneficial to have at the top of the file (maybe in the general function section I suggested or brought up in [#232 (comment)](https://github.com/ioccc-src/mkiocccentry/issues/232#issuecomment-1146792747) ?) an example which has all functions in use? I'm not sure: it might be too much for some people but it might also give an entire overview of the facility for people to see.
> 
> I'm not really sure here.

Probably a summary of function types near the top would be a good idea, maybe even in the `TL;DR` section?  Maybe a simplified recap of the external function prototypes (without those complex `__attribute__` stuff) in `dbg.h` with links down to the sections where the details of the functions are discussed.  This might allow you to combine the table of contents and the  `TL;DR` int one section.

**Comment by lcn2** [comment]
> Again these are VERY loose ideas! I'm also sorry for anything that might not be worded so well or not as clear. I'm going to take a break real soon and I hope that with some discussion I can start working on this tomorrow.

Probably best to try the idea and see how it looks.

**Comment by lcn2** [comment]
> One more and then I am taking a break for a while. (Make that two .. or three depending on if you want to count man page, another file entirely, as another point: we might want to bring up the `va_list` versions briefly at the top but point out that they're not recommended - for the reasons given).
> 
> ## Output control
> It might be an idea to move this towards the top too? It would kind of be a merge of the different ones. It might be in the synopses of each?
> 
> Might I also suggest we write (I am happy to do this) a man page? That could be a way to make it easier to see how to use the functions. Would it be section 3? I'd guess that but maybe you have another idea?
> 
> Going to take a break now. Will discuss these things with you and go from there. Again sorry for any inconsistencies or for any text that's not clear and even the poor examples. I just wanted to get the ideas out to discuss quickly so I could go rest.

There are two cases where output control info is useful, in a specific function section where someone wants to know how to control that function (which is the way `dbg.md` does it now) AND when someone wants to see the complete dbg facility controls, where is what the _global message control variables_ section in the current `dbg.md`  document was intended to address, but perhaps needs to be completed.

**Comment by xexyl** [comment]
> > ## TL; DR section
> > I moved the error details to this section that were mostly duplicated across all the sections. There was an exception though where it depended on a boolean: that one I left there. I guess that was `warn_or_err` but maybe not.
> > As of commit [249f725](https://github.com/ioccc-src/mkiocccentry/commit/249f72541c04dfe11671a27a904d50f525ffd8c6) I did the same for printf formats. I think this type of thing will help make it easier to read. There might be some cases where this is not the case though including ironically if someone skips past the `tl;dr` section and goes straight to the functions they're after.
> > This makes me wonder if there should be a general synopsis section at the top as well. It might discuss the different capabilities of the facility and also have all the function prototypes (perhaps without the comments at the top). Then each prototype could be a link to the actual function details. Perhaps a back link would allow them to go back to this section.
> > That might make it easier to find the different facilities.
> > What do you think here?
> 
> What is moved into that section should be kept to a minimum as making the `TL;DR` too long kind of defeats the purpose. :-)

Quite true indeed.

> 
> Moreover the `TL;DR` section should be a very short summary. Perhaps it is better for the `TL;DR` section might refer to some _common error section_ and simply and briefly mention that if the user runs into problems with a link to that _common error section_?

Would you be referring to for instance:

```md
# General notes for all functions

## `fmt` argument

The `fmt` argument in the functions is a printf-style format. If the format
requires arguments, then such arguments may be given after the `fmt`.

[...]

### In case of NULL `fmt`:

If `fmt == NULL`, then the following format string will be used:

```
((NULL fmt))
```

and the following warning, preceded by a newline, will be issued:

[...]

## In case of NULL name

In the functions that accept a name (say the calling function) if `name ==
NULL`, then the following string will be used instead:

```
((NULL name))
```
[...]

```

? Because actually I would say that's no longer in the `tl;dr` section. Perhaps there should be a reference to this though?

I would also think that it should refer to the man page as I think that could be compact and once done it might even be a good layout for the `tl;dr` section. What do you think? Maybe I'll do the man page next instead of table of contents.

I'm afraid that's most likely not going to happen today though. Sorry :( It's been a difficult day so far getting to do anything at all though I at least got more test cases for the `txzchk` test suite which is good. I think that's all that's needed for now in that area.

**Comment by xexyl** [comment]
> > ## Examples at the top of the file
> > Would it be beneficial to have at the top of the file (maybe in the general function section I suggested or brought up in [#232 (comment)](https://github.com/ioccc-src/mkiocccentry/issues/232#issuecomment-1146792747) ?) an example which has all functions in use? I'm not sure: it might be too much for some people but it might also give an entire overview of the facility for people to see.
> > I'm not really sure here.
> 
> Probably a summary of function types near the top would be a good idea, maybe even in the `TL;DR` section? Maybe a simplified recap of the external function prototypes (without those complex `__attribute__` stuff) in `dbg.h` with links down to the sections where the details of the functions are discussed. This might allow you to combine the table of contents and the `TL;DR` int one section.

Probably in the `tl;dr` section yes because they at least need to read that (though some might look at the code instead I guess). 

Links are a good idea yes.

It might do what you suggest it might do though I guess we'll have to see. I will have to use discount to generate the html file to test - locally - but that's okay and I tend to do that when editing markdown documents anyway. I think I actually will have to do this in order to proceed though as noted in the above comment I think writing a man page might be a good start.

Which section should it be? 3? That would be of help (though we could always change it to be a different section of course).

**Comment by xexyl** [comment]
> > Again these are VERY loose ideas! I'm also sorry for anything that might not be worded so well or not as clear. I'm going to take a break real soon and I hope that with some discussion I can start working on this tomorrow.
> 
> Probably best to try the idea and see how it looks.

Going back to my first reply: this is I think something I should also do for the table of contents: try and see how it looks. 

I think though I need a clearer head in order to do this which means today I won't get anything done here. Sorry! Depending on tonight hopefully in the morning I can work on this. Maybe later on I can start a man page but I don't think that's going to happen today. I still do think though that since man pages can be compact it might be a good basis for the `tl;dr` section. I'll see if I can recall some man pages which describe how to set up some code or else find them (maybe with grep). Maybe even `man man-pages` will be of help. I'll worry about that next.

Maybe once I've made a man page you can give your thoughts on it and then I could go from there as far as setting up `dbg.md` ? Does that sound like a good idea?

Going afk again for a while but when I'm back I will be doing something else. 

Good day!

**Comment by lcn2** [comment]
> > > ## Hyperlinks
> > > As I suggested the other day I think this would help a lot. But what sort of hyperlinks? Obviously a table of contents. I would think that each section would also have a short list of links as well: for example at the top of each section it would allow one to jump to the respective function.
> > > Another thought: should each reference to a function (outside of code blocks) be a link to the function prototype? I could see how this might be an issue but it might also be helpful to some people.
> > > What do you think? If I think of other things with links when going through the document I'll bring them up in another comment.
> > 
> > 
> > Probably a compact table of contents so that someone can quickly jump down into the middle of the document would be the most useful.
> 
> Sounds good. What should the items in the table be though? That's something that has to be worked out too. This in a way related to another comment of yours so I'll delay the rest for that comment.

Perhaps a link per function set such as:

- `err()`
    ... and related functions
- `errp()`
    ... and related functions
- `werr()`
    ... and related functions
- `werrp()`
    ... and related functions
- `warn_or_err()`
    ... and related functions
- `warnp_or_errp()`
    ... and related functions
- `printf_usage()`
    ... and related functions

Maybe?

We don't think that every _dbg facility_ function needs a separate section.  On the other hand the _p_ versions are different enough to warrant being in a separate section.

**Comment by lcn2** [comment]
> I would also think that it should refer to the man page as I think that could be compact and once done it might even be a good layout for the tl;dr section. What do you think? Maybe I'll do the man page next instead of table of contents.

Yes, referring to a `dbg(3)` man page, and adding one, if a good idea.

**Comment by lcn2** [comment]
> > > ## Examples at the top of the file
> > > Would it be beneficial to have at the top of the file (maybe in the general function section I suggested or brought up in [#232 (comment)](https://github.com/ioccc-src/mkiocccentry/issues/232#issuecomment-1146792747) ?) an example which has all functions in use? I'm not sure: it might be too much for some people but it might also give an entire overview of the facility for people to see.
> > > I'm not really sure here.
> > 
> > 
> > Probably a summary of function types near the top would be a good idea, maybe even in the `TL;DR` section? Maybe a simplified recap of the external function prototypes (without those complex `__attribute__` stuff) in `dbg.h` with links down to the sections where the details of the functions are discussed. This might allow you to combine the table of contents and the `TL;DR` int one section.
> 
> Probably in the `tl;dr` section yes because they at least need to read that (though some might look at the code instead I guess).
> 
> Links are a good idea yes.
> 
> It might do what you suggest it might do though I guess we'll have to see. I will have to use discount to generate the html file to test - locally - but that's okay and I tend to do that when editing markdown documents anyway. I think I actually will have to do this in order to proceed though as noted in the above comment I think writing a man page might be a good start.
> 
> Which section should it be? 3? That would be of help (though we could always change it to be a different section of course).

3 seems good for now.

**Comment by lcn2** [comment]
> > > Again these are VERY loose ideas! I'm also sorry for anything that might not be worded so well or not as clear. I'm going to take a break real soon and I hope that with some discussion I can start working on this tomorrow.
> > 
> > 
> > Probably best to try the idea and see how it looks.
> 
> Going back to my first reply: this is I think something I should also do for the table of contents: try and see how it looks.
> 
> I think though I need a clearer head in order to do this which means today I won't get anything done here. Sorry! Depending on tonight hopefully in the morning I can work on this. Maybe later on I can start a man page but I don't think that's going to happen today. I still do think though that since man pages can be compact it might be a good basis for the `tl;dr` section. I'll see if I can recall some man pages which describe how to set up some code or else find them (maybe with grep). Maybe even `man man-pages` will be of help. I'll worry about that next.
> 
> Maybe once I've made a man page you can give your thoughts on it and then I could go from there as far as setting up `dbg.md` ? Does that sound like a good idea?

Yes, a `dbd(3)` man page is needed.  This would allow the`dbg.md` file to be simplified.

**Comment by xexyl** [comment]
> > > > ## Hyperlinks
> > > > As I suggested the other day I think this would help a lot. But what sort of hyperlinks? Obviously a table of contents. I would think that each section would also have a short list of links as well: for example at the top of each section it would allow one to jump to the respective function.
> > > > Another thought: should each reference to a function (outside of code blocks) be a link to the function prototype? I could see how this might be an issue but it might also be helpful to some people.
> > > > What do you think? If I think of other things with links when going through the document I'll bring them up in another comment.
> > > 
> > > 
> > > Probably a compact table of contents so that someone can quickly jump down into the middle of the document would be the most useful.
> > 
> > 
> > Sounds good. What should the items in the table be though? That's something that has to be worked out too. This in a way related to another comment of yours so I'll delay the rest for that comment.
> 
> Perhaps a link per function set such as:
> 
> * `err()`
>   ... and related functions
> * `errp()`
>   ... and related functions
> * `werr()`
>   ... and related functions
> * `werrp()`
>   ... and related functions
> * `warn_or_err()`
>   ... and related functions
> * `warnp_or_errp()`
>   ... and related functions
> * `printf_usage()`
>   ... and related functions
> 
> Maybe?
> 
> We don't think that every _dbg facility_ function needs a separate section. On the other hand the _p_ versions are different enough to warrant being in a separate section.


These sound reasonable. I'm not sure if the _p_ versions should be with their respective function(s). I think that would be the better way to go about it.

**Comment by xexyl** [comment]
> > I would also think that it should refer to the man page as I think that could be compact and once done it might even be a good layout for the tl;dr section. What do you think? Maybe I'll do the man page next instead of table of contents.
> 
> Yes, referring to a `dbg(3)` man page, and adding one, if a good idea.



> Yes, a `dbd(3)` man page is needed. This would allow the`dbg.md` file to be simplified.

Agreed on the simplifying the `dbg.md` file as well. Perhaps much of the `tl;dr` section could be removed even. I'm not sure of the audience so it might not be that all of it can be removed - I'll let you decide that - but we can at least simplify it some. I think I'll work on this soon and then do a pull request.

I'll refer to it in the markdown file as well but whether I do anything else there today or not I do not know. I hope to have this done in a bit but we'll see: I don't know yet how today will be going.

**Comment by xexyl** [comment]
See commit 881b1970a8c291180693d9e1d6b86552eaf476ec:

    Initial dbg.3 man page
    
    This is not likely complete but it's good start for initial analysis and
    discussion.
    
    More examples should be added (from each section in the dbg.md file).
    
    Formatting has to be checked in its entirety. An example is in C code.
    Instead of having - one has to have \- (even in C code). I believe I got
    all cases but I might have missed some. Similarly it appears that \n in
    C should be in the man page \\n.
    
    Consistency should be checked more thoroughly as well. A good example is
    making sure that the va_list functions are in the same order as the
    regular functions when indicating equivalency (other than taking a
    va_list). Another example of consistency that needs to be checked is the
    order of the functions in the man page versus dbg.h.
    
    The dbg.md does not yet refer to the man page and won't until it is
    completed and made as concise and precise as possible.

It would be good if we could discuss all these things perhaps after you've looked at the man page.

**Comment by xexyl** [comment]
Also:

    Something I neglected to mention in commit
    881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
    about compiling dbg.c and linking dbg.o into the binaries that wish to
    use the dbg facility. Where and how this should be displayed is not
    something I am sure of yet.

**Comment by lcn2** [comment]
> See commit [881b197](https://github.com/ioccc-src/mkiocccentry/commit/881b1970a8c291180693d9e1d6b86552eaf476ec):
> 
> ```
> Initial dbg.3 man page
> 
> This is not likely complete but it's good start for initial analysis and
> discussion.
> 
> More examples should be added (from each section in the dbg.md file).
> 
> Formatting has to be checked in its entirety. An example is in C code.
> Instead of having - one has to have \- (even in C code). I believe I got
> all cases but I might have missed some. Similarly it appears that \n in
> C should be in the man page \\n.
> 
> Consistency should be checked more thoroughly as well. A good example is
> making sure that the va_list functions are in the same order as the
> regular functions when indicating equivalency (other than taking a
> va_list). Another example of consistency that needs to be checked is the
> order of the functions in the man page versus dbg.h.
> 
> The dbg.md does not yet refer to the man page and won't until it is
> completed and made as concise and precise as possible.
> ```
> 
> It would be good if we could discuss all these things perhaps after you've looked at the man page.

A quick look reading the nroff looks good.  We might add a `.SH HISTORY` section once the man page exists.

**Comment by xexyl** [comment]
> > See commit [881b197](https://github.com/ioccc-src/mkiocccentry/commit/881b1970a8c291180693d9e1d6b86552eaf476ec):
> > ```
> > Initial dbg.3 man page
> > 
> > This is not likely complete but it's good start for initial analysis and
> > discussion.
> > 
> > More examples should be added (from each section in the dbg.md file).
> > 
> > Formatting has to be checked in its entirety. An example is in C code.
> > Instead of having - one has to have \- (even in C code). I believe I got
> > all cases but I might have missed some. Similarly it appears that \n in
> > C should be in the man page \\n.
> > 
> > Consistency should be checked more thoroughly as well. A good example is
> > making sure that the va_list functions are in the same order as the
> > regular functions when indicating equivalency (other than taking a
> > va_list). Another example of consistency that needs to be checked is the
> > order of the functions in the man page versus dbg.h.
> > 
> > The dbg.md does not yet refer to the man page and won't until it is
> > completed and made as concise and precise as possible.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It would be good if we could discuss all these things perhaps after you've looked at the man page.
> 
> A quick look reading the nroff looks good. We might add a `.SH HISTORY` section once the man page exists.

The quick look wrt the nroff .... you mean the formatting of e.g. `-` (`\-`) etc.?

And yes I think a HISTORY section might be an idea. I still think it could use more work though. I can add more to it tomorrow.

The question of course is once that's done what should be changed in `dbg.md`. I guess we have to wait and see but if you do have any comments once you've actually used `man` to generate the man page I'm open to them.

**Comment by lcn2** [comment]
> Also:
> 
> ```
> Something I neglected to mention in commit
> 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> about compiling dbg.c and linking dbg.o into the binaries that wish to
> use the dbg facility. Where and how this should be displayed is not
> something I am sure of yet.
> ```

Perhaps a simple example in both the man age and the markdown document is in order?

The `main()` function in `dbg.c` is a bit too complex.    Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.

One could mention that for a more extended example, the reader should look at  `main()` function in `dbg.c` as well.

**Comment by xexyl** [comment]
> > The quick look wrt the nroff .... you mean the formatting of e.g. - (-) etc.?
> 
> We mean reading the nroff code without formatting it. :-)

Right. I got that. I was more wondering if the format looks good with the concern I brought up (or maybe not concern but a point to make sure it is okay with the escaped chars). But perhaps I took that too literally and maybe you were considering it in its entirety.

**Comment by xexyl** [comment]
> > Also:
> > ```
> > Something I neglected to mention in commit
> > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > use the dbg facility. Where and how this should be displayed is not
> > something I am sure of yet.
> > ```
> 
> Perhaps a simple example in both the man age and the markdown document is in order?
> 

You mean an example program?

> The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.

Would it also parse verbosity level?

> 
> One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.

True.

**Comment by lcn2** [comment]
> > > Also:
> > > ```
> > > Something I neglected to mention in commit
> > > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > > use the dbg facility. Where and how this should be displayed is not
> > > something I am sure of yet.
> > > ```
> > 
> > 
> > Perhaps a simple example in both the man age and the markdown document is in order?
> 
> You mean an example program?
> 
> > The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.
> 
> Would it also parse verbosity level?
> 
> > One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.
> 
> True.

To keep is simple, perhaps just assign verbosity_level as a constant and mention:

```c
/* This is just a trivial demo, see the main function in dbg.c for a better example */

...

{
    ...
    verbosity_level = DBG_MED;    /* we suggest you use getopt(3) convert -v level on command line into this value */
    ...
}
```

**Comment by xexyl** [comment]
> > > > Also:
> > > > ```
> > > > Something I neglected to mention in commit
> > > > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > > > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > > > use the dbg facility. Where and how this should be displayed is not
> > > > something I am sure of yet.
> > > > ```
> > > 
> > > 
> > > Perhaps a simple example in both the man age and the markdown document is in order?
> > 
> > 
> > You mean an example program?
> > > The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.
> > 
> > 
> > Would it also parse verbosity level?
> > > One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.
> > 
> > 
> > True.
> 
> To keep is simple, perhaps just assign verbosity_level as a constant and mention:
> 
> ```c
> /* This is just a trivial demo, see the main function in dbg.c for a better example */
> 
> ...
> 
> {
>     ...
>     verbosity_level = DBG_MED;    /* we suggest you use getopt(3) convert -v level on command line into this value */
>     ...
> }
> ```

Good idea.

So the simple example in both `dbg.md` and `dbg.3` would be a complete program that:

- assigns a value to verbosity_level
- makes a few calls to some of the functions

There would be some calls that would not show output (indicate it in comments) due to verbosity level not being high enough and some that would output text.

Does that sound about right?

Do you want to wait on the discussion of how to modify the md file otherwise until the man page is completed ?

**Comment by lcn2** [comment]
> > > > > Also:
> > > > > ```
> > > > > Something I neglected to mention in commit
> > > > > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > > > > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > > > > use the dbg facility. Where and how this should be displayed is not
> > > > > something I am sure of yet.
> > > > > ```
> > > > 
> > > > 
> > > > Perhaps a simple example in both the man age and the markdown document is in order?
> > > 
> > > 
> > > You mean an example program?
> > > > The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.
> > > 
> > > 
> > > Would it also parse verbosity level?
> > > > One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.
> > > 
> > > 
> > > True.
> > 
> > 
> > To keep is simple, perhaps just assign verbosity_level as a constant and mention:
> > ```c
> > /* This is just a trivial demo, see the main function in dbg.c for a better example */
> > 
> > ...
> > 
> > {
> >     ...
> >     verbosity_level = DBG_MED;    /* we suggest you use getopt(3) convert -v level on command line into this value */
> >     ...
> > }
> > ```
> 
> Good idea.
> 
> So the simple example in both `dbg.md` and `dbg.3` would be a complete program that:
> 
> * assigns a value to verbosity_level
> * makes a few calls to some of the functions
> 
> There would be some calls that would not show output (indicate it in comments) due to verbosity level not being high enough and some that would output text.
> 
> Does that sound about right?

That sounds reasonable.

**Comment by lcn2** [comment]
> Do you want to wait on the discussion of how to modify the md file otherwise until the man page is completed ?

Sure.

We are going back to bug fixing ...

**Comment by xexyl** [comment]
> > > > > > Also:
> > > > > > ```
> > > > > > Something I neglected to mention in commit
> > > > > > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > > > > > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > > > > > use the dbg facility. Where and how this should be displayed is not
> > > > > > something I am sure of yet.
> > > > > > ```
> > > > > 
> > > > > 
> > > > > Perhaps a simple example in both the man age and the markdown document is in order?
> > > > 
> > > > 
> > > > You mean an example program?
> > > > > The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.
> > > > 
> > > > 
> > > > Would it also parse verbosity level?
> > > > > One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.
> > > > 
> > > > 
> > > > True.
> > > 
> > > 
> > > To keep is simple, perhaps just assign verbosity_level as a constant and mention:
> > > ```c
> > > /* This is just a trivial demo, see the main function in dbg.c for a better example */
> > > 
> > > ...
> > > 
> > > {
> > >     ...
> > >     verbosity_level = DBG_MED;    /* we suggest you use getopt(3) convert -v level on command line into this value */
> > >     ...
> > > }
> > > ```
> > 
> > 
> > Good idea.
> > So the simple example in both `dbg.md` and `dbg.3` would be a complete program that:
> > 
> > * assigns a value to verbosity_level
> > * makes a few calls to some of the functions
> > 
> > There would be some calls that would not show output (indicate it in comments) due to verbosity level not being high enough and some that would output text.
> > Does that sound about right?
> 
> That sounds reasonable.

Sounds good to me then. 

I will look at it tomorrow. I do have a doctor appointment tomorrow and also Friday but most likely I will be awake very early again anyway so presumably I will be able to get some of this done.

**Comment by xexyl** [comment]
> > Do you want to wait on the discussion of how to modify the md file otherwise until the man page is completed ?
> 
> Sure.
> 
> We are going back to bug fixing ...

Enjoy! I don’t mean that cynically mind you. I find bug finding and fixing to be quite enjoyable. If you don’t though I hope it goes well and you can resolve it quickly (even if it is enjoyable though that’s probably better so that you can focus on the things that haven’t been worked on recently). 

I have been having difficult days recently but I am thankfully able to work on the things here. It gives me a purpose and I enjoy helping as well. 

Because it’s been difficult (and really slow) lately I don’t know how long this will take but I hope that tomorrow I can finish the man page. After that we can discuss how to integrate it into the markdown file. 

Once that is discussed more I can put you together and after that we can do a couple passes to see that everything is in order. Once they are I will let you know that it looks good on my end. I guess I could always do a pull request on the new repo as well though. 

I almost added a BUGS section to the man page today but since the repo doesn’t exist yet I didn’t add it. I guess you can do that once the URI is figured out. 

I hope you have a great rest of your day and sleep well when you do! I will be at the phone for a bit more but then I will be away for many hours.

**Comment by xexyl** [comment]
Oh if you haven’t looked at the uuencode issue after this is done maybe I will consider looking into it as well. I believe that this is a higher priority though. 

Perhaps it’s the highest priority if everything here even?

Is there something that is actually more important right now? I don’t mind. I just want to be sure as it occurred to me that possibly there are other things that might be more important in your mind. 

I am not sure about the json parser right now but I have gathered that you have some ideas on how to proceed and since I have been working on other things that are also important I have kind of let that be for now. 

Anyway if you prefer something else is worked on please let me know.

**Comment by lcn2** [comment]
> Oh if you haven’t looked at the uuencode issue after this is done maybe I will consider looking into it as well. I believe that this is a higher priority though.

For stuff that is blocked .. that may be a good thing to work on in parallel with the _dbg facility_ documentation.

> I am not sure about the json parser right now but I have gathered that you have some ideas on how to proceed and since I have been working on other things that are also important I have kind of let that be for now.

We have a general idea of how to proceed with the parser, but we should fix issue #214 first.

**Comment by lcn2** [comment]
> I will look at it tomorrow. I do have a doctor appointment tomorrow and also Friday but most likely I will be awake very early again anyway so presumably I will be able to get some of this done.

Best wishes on those appointments.

**Comment by xexyl** [comment]
> > I will look at it tomorrow. I do have a doctor appointment tomorrow and also Friday but most likely I will be awake very early again anyway so presumably I will be able to get some of this done.
> 
> Best wishes on those appointments.

Thank you!

**Comment by xexyl** [comment]
> > Oh if you haven’t looked at the uuencode issue after this is done maybe I will consider looking into it as well. I believe that this is a higher priority though.
> 
> For stuff that is blocked .. that may be a good thing to work on in parallel with the _dbg facility_ documentation.
> 

What do you consider blockers right now please?

> > I am not sure about the json parser right now but I have gathered that you have some ideas on how to proceed and since I have been working on other things that are also important I have kind of let that be for now.
> 
> We have a general idea of how to proceed with the parser, but we should fix issue #214 first.

That sounds great! And after that I can start working on jinfochk and jauthchk as well. 

I am not sure if I replied to everything yet. I think there’s at least one that I haven’t and maybe there will be more when I look again but I am heading off now for sleep. 

Good night my friend!

**Comment by xexyl** [comment]
> > > > > Also:
> > > > > ```
> > > > > Something I neglected to mention in commit
> > > > > 881b1970a8c291180693d9e1d6b86552eaf476ec is that there's no explanation
> > > > > about compiling dbg.c and linking dbg.o into the binaries that wish to
> > > > > use the dbg facility. Where and how this should be displayed is not
> > > > > something I am sure of yet.
> > > > > ```
> > > > 
> > > > 
> > > > Perhaps a simple example in both the man age and the markdown document is in order?
> > > 
> > > 
> > > You mean an example program?
> > > > The `main()` function in `dbg.c` is a bit too complex. Perhaps an example that is calls `dbg()` and `err()` and/or `errp()` could be added.
> > > 
> > > 
> > > Would it also parse verbosity level?
> > > > One could mention that for a more extended example, the reader should look at `main()` function in `dbg.c` as well.
> > > 
> > > 
> > > True.
> > 
> > 
> > To keep is simple, perhaps just assign verbosity_level as a constant and mention:
> > ```c
> > /* This is just a trivial demo, see the main function in dbg.c for a better example */
> > 
> > ...
> > 
> > {
> >     ...
> >     verbosity_level = DBG_MED;    /* we suggest you use getopt(3) convert -v level on command line into this value */
> >     ...
> > }
> > ```
> 
> Good idea.
> 
> So the simple example in both `dbg.md` and `dbg.3` would be a complete program that:
> 
> * assigns a value to verbosity_level
> * makes a few calls to some of the functions
> 
> There would be some calls that would not show output (indicate it in comments) due to verbosity level not being high enough and some that would output text.
> 
> Does that sound about right?
> 
> Do you want to wait on the discussion of how to modify the md file otherwise until the man page is completed ?

Oh quickly before I put the phone down. 

How do you propose that the compilation steps should be listed?

I actually wonder if a basic Makefile would be if use. I don’t know but maybe something could be contrived. 

Well sleep time for me. Good night and sleep well when you do!

**Comment by lcn2** [comment]
> What do you consider blockers right now please?

The code needed to parse compound JSON items. We have an idea of how to do this now.

**Comment by lcn2** [comment]
> How do you propose that the compilation steps should be listed?

Just show a few simple shell lines that relate to the example.

**Comment by xexyl** [comment]
> > What do you consider blockers right now please?
> 
> The code needed to parse compound JSON items. We have an idea of how to do this now.

Great!

**Comment by xexyl** [comment]
> > How do you propose that the compilation steps should be listed?
> 
> Just show a few simple shell lines that relate to the example.

Okay. I have to think on the best presentation (though I don't think the initial will be the best one .. there are several ways that might be possible). I also have to think of where to put it. The synopsis section doesn't seem right in one way but in another way I could see it being the right section: not sure yet.

But then again maybe it can go along the lines of how you have to link some libraries in for some functions. These are I think usually (maybe not always but usually) in the SYNOPSIS section. If so that might be a place to include it but again not sure.

I hope to look at this soon.

**Comment by xexyl** [comment]
I have a fun example along with how to compile it and the dbg facility and link it into the binary. Then it runs it. 

It doesn’t do any tests or conversions. For `errp` it just explicitly sets `errno`  Each part has comments to explain it and then the output is shown as well. 

I actually have a test file that I will commit that has the same contents. I will either add a Makefile for this or add a rule to the main one that you could move over to the new repo. I might make it so it has object files that the user can update for easier compilation. That would necessitate a different Makefile. I will add the test file but whether I add the Makefile or not I don’t know yet. Maybe I will hold off on that to see what you think. 

Yes I think that. So the man page does cat on the file (so that they can see the contents) and then runs cc to first compile the dbg facility and the test file and then again to link them into a binary. Then finally it will run it.

In fact this is already done. I think though that I will discuss where it belongs in the markdown file after you see it. 

I am going to take a break and hopefully I can commit and make a pull request today. It largely depends on how I feel after my break. I will be back after my doctor appointment (it’s just blood work) and I will possibly be able to do more if I haven’t done these things yet but we shall see. I am very very tired right now but I think this example will be of good value. 

Hope you’re having a nice sleep my friend!

**Comment by xexyl** [comment]
See commit https://github.com/ioccc-src/mkiocccentry/commit/c482f0fdcfe0f9b2d0bfb7fd9cf7c725defd076a:

    First complete pass of dbg.3 man page
    
    The man page is now ready for a thorough review and if everything checks
    out okay the dbg.md file could be updated to refer to it. It's possible
    another pass over the man page would be a good idea though. This
    specifically goes for what I noted in commit
    881b1970a8c291180693d9e1d6b86552eaf476ec. I believe I can tick these but
    they should still be checked nonetheless:
    
        Formatting has to be checked in its entirety. An example is in C code.
        Instead of having - one has to have \- (even in C code). I believe I got
        all cases but I might have missed some. Similarly it appears that \n in
        C should be in the man page \\n.
    
    Note that in this case the dbg_example.c file itself obviously cannot
    have the '\'s in front of '-'s and '\'s.
    
    What I cannot say can be ticked yet is:
    
        Consistency should be checked more thoroughly as well. A good example is
        making sure that the va_list functions are in the same order as the
        regular functions when indicating equivalency (other than taking a
        va_list). Another example of consistency that needs to be checked is the
        order of the functions in the man page versus dbg.h.
    
    For the first part I have not done it at all though for writing it in
    the first place I made each function on its own line and then yanked and
    pasted and then removed the v from the function name so it should be
    okay but I still think it should be checked again.
    
    For the second part I believe I originally had it as well but I could be
    wrong as I did go about doing it in different orders at first.
    
    What can now likely be discussed is this:
    
        The dbg.md does not yet refer to the man page and won't until it is
        completed and made as concise and precise as possible.
    
    Though it might be better to make sure the man page is finished first.
    Perhaps not though as reducing the text in dbg.md and referring to the
    man page might actually help.
    
    The example in the man page now is a complete program (see
    dbg_example.c) and it shows the contents (via cat) and then how to
    compile it along with dbg.c. After this it runs './dbg_example' and
    shows the output. Finally it shows the exit status of the dbg_example
    (which will actually be 2 as I made the final line an errp() call).
    
    For reference this is how to compile the dbg_example:
    
            $ cc -c dbg.c
            $ cc -o dbg_example dbg_example.c dbg.o
    
    Alternatively you can do:
    
            $ cc -o dbg_example dbg_example.c dbg.c
    
    A Makefile could have e.g.:
    
        dbg_example: dbg.o dbg_example.c
                ${CC} ${CFLAGS} dbg.o dbg_example.c -o $@
    
    or
    
        dbg_example: dbg.c dbg.h dbg_example.c
                ${CC} ${CFLAGS} dbg.c dbg_example.c -o $@
    
    for two examples (this list is definitely not exhaustive).


I believe this is a good place to start discussing how to update `dbg.md` but also discuss `dbg.3` as well.

Taking a break now and I don't expect to return to working on the files later on today. I can however discuss the changes and what needs to be done still.

**Comment by xexyl** [comment]
Leaving for the day but wanted to bring up a couple things.

First is that we still have to discuss this issue now that the man page is in good condition to be analysed.

Second is OT for this issue but I want to say it quickly so I can head off. I'm aware of an issue with some bad files in the bad directory for the txzchk test script. Here's what I mean:

```sh
$ grep 'tar fvt' *
entry.test-0.1652598667.txt:tar fvt entry.test-0.1644411114.txz
entry.test-2.1652598667.txt:tar fvt entry.test-2.1644411114.txz
entry.test-3.1652598667.txt:tar fvt entry.test-3.1644411114.txz
entry.test-4.1652598667.txt:tar fvt entry.test-4.1644411114.txz
entry.test-5.1652598667.txt:tar fvt entry.test-5.1644411114.txz
```

These were accidentally committed and I'm not sure how except that I was making the text files from tarballs I actually had - that's all I can be sure of. I should still have these tarballs so I will tomorrow look at them (if I can - busier day tomorrow and that includes a doctor appointment but it's not the only thing that has to be done). I'm not sure if their contents are already in another text file or not and it might not be needed to have them even if they're not but I'm too tired to take care of it tonight.

Hope you have a great rest of your day and sleep well when you do my friend! Look forward to discussing this issue though.

**Comment by xexyl** [comment]
Oh another thing. Today the dbg_example rule was committed in the pull request I made. The .gitignore file already was updated though.

Anyway good night!

**Comment by lcn2** [comment]
Perhaps the documentation is ready for an initial release?  Or should we wait?

**Comment by lcn2** [comment]
In the next comment we will details about discuss commit 4b356b3ba9e739532573153268a4829b05a2a3d7

    Released dbg 2.2 2022-06-08

    Changed dbg facility version to 2.2 2022-06-08.

    Added functions to determine if certain debug messages are allowed:

        extern bool msg_allowed(void);
        extern bool dbg_allowed(int level);
        extern bool warn_allowed(void);
        extern bool err_allowed(void);
        extern bool usage_allowed(void);

    Now all dbg() functions use the above functions to determine if
    messages are allowed to be written.

    Probably the most useful is to determine if a given debug level
    message is allowed.  For example, say there a complex
    debugging action should only be done if the debug level
    is the right level, instead of making an explicit check
    against the verbosity_level global, a simple call to
    to dbg_allowed is needed:

    for (i=0; i < bigvalue; ++i) {

        ...

        if (dbg_allowed(DBG_VVHIGH)) {
            dbg(DBG_VVHIGH, "start of special debug section")l
            dbg(DBG_VVHIGH, "i: %ld a: %s b: %d c: %f",
                            i, intensive_a_func(i), intensive_b_func(i), intensive_c_func(i));
            (void) fprintf(dbg_stream, "special message: %s\n", special(i));
            more_complex_debugging(i);
            (void) fprintf(dbg_stream, "very special message: %s\n", very_special(i));
            dbg(DBG_VVHIGH, "end of special debug section")l
        }

        ...
    }

    By pre-checking if the debug level DBG_VVHIGH is active, calls
    to complex debugging functions and even special writes outside
    of the dbg facility can be avoided.

**Comment by lcn2** [comment]
Regarding commit 4b356b3ba9e739532573153268a4829b05a2a3d7, we did not update `dbg.3` nor `dbg.md` because we didn't know if you have some plans to edits in the works where modifying those files would complicate your efforts.

We decided we needed a common way to check if certain dbg facility messages are allowed by providing a simple boolean returning function.  Perhaps as you improve the documentation, you might want to reference such functions:

```c
extern bool msg_allowed(void);
extern bool dbg_allowed(int level);
extern bool warn_allowed(void);
extern bool err_allowed(void);
extern bool usage_allowed(void);
```

We plan to write a equivalent message for the JSON debug facility.  For example, the `json_dbg_allowed(level)` function can be used to surround some of the existing JSON parser debug writes (even ones that write without calling `json_dbg()`).

Sorry (tm Canada 🇨🇦) to edit and run .. we are in the process of trying to get the `jparse` diagnostics ready for the complex and compound JSON item parsing.

**Comment by lcn2** [comment]
Released dbg 2.3 2022-06-16 to a new [GitHub repo](https://github.com/lcn2/dbg).

**Comment by lcn2** [comment]
With commit eb93987bf998f4168a80105a8b0f397f806fb0f2, synced with [dbg GitHub repo](https://github.com/lcn2/dbg).

**Comment by xexyl** [comment]
Just so you know I was not done with documentation but we can discuss it here if you like and I can do a pull request there. 

Or if you’re satisfied with the documentation then that’s okay too. 

Hope to reply to the other comments tomorrow.

**Comment by lcn2** [comment]
> Just so you know I was not done with documentation but we can discuss it here if you like and I can do a pull request there.
> 
> Or if you’re satisfied with the documentation then that’s okay too.
> 
> Hope to reply to the other comments tomorrow.

Opps, sorry (tm Canada 🇨🇦).

Well, what is done is done.

We guess the "not done with documentation" can form a version 2.5 release.

**Comment by xexyl** [comment]
> > Just so you know I was not done with documentation but we can discuss it here if you like and I can do a pull request there.
> > Or if you’re satisfied with the documentation then that’s okay too.
> > Hope to reply to the other comments tomorrow.
> 
> Opps, sorry (tm Canada 🇨🇦) @xexyl.
> 
> Well, what is done is done.
> 
> We guess the "not done with documentation" can form a version 2.5 release.
> 

However you prefer. Should we discuss it here and I make changes there? Or will you copy them over?

> ### UPDATE 0:
> Reopening this issue.

Sure. What do you think of the man page as it stands? I don’t have the laptop out now and won’t until tomorrow but that’s what I was holding off for (though this week I have been incredibly tired which doesn’t help). 

I think if we can figure out what should change with the man page we can then go from there and figure out how to update the markdown file. 

Assuming you get to this before tomorrow I will hopefully be able to reply. 

Hope you have a great night!

**Comment by lcn2** [comment]
> However you prefer. Should we discuss it here and I make changes there? Or will you copy them over?

We suggest you could form an issue on the [dbg repo](https://github.com/lcn2/dbg) and work with pull requests over in that repo.

Later on, when there is a useful dbg release, files could be copied over into this repo.

There we will again close this issue.
